core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -87,13 +90,71 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
 
   /**
    * Registers a given broker in zookeeper as the controller.
+   * @return the updated controller epoch and epoch zkVersion
    * @param controllerId the id of the broker that is to be registered as the controller.
    * @param timestamp the timestamp of the controller election.
-   * @throws KeeperException if an error is returned by ZooKeeper.
-   */
-  def registerController(controllerId: Int, timestamp: Long): Unit = {
-    val path = ControllerZNode.path
-    checkedEphemeralCreate(path, ControllerZNode.encode(controllerId, timestamp))
+   * @throws ControllerMovedException if fail to create /controller or fail to increment controller epoch.
+   */
+  def registerController(controllerId: Int, timestamp: Long): (Int, Int) = {
+    getControllerEpoch match {
```

Comment:
No. 
Previously we create `/controller_epoch` on-demand if it does not exist when we try to increment controller epoch. IMO, this makes the code hard to read and reason about, especially after this patch because in that way we need to either create `\controller_epoch` if not exists and retry the atomic operation or we have two different atomic operations (one for check+create, the other one for create+create). 

I think `\controller_epoch` should pre-exists before we actually use it, like other persistent zk paths (e.g. /brokers, /admin/delete_topics) . So I have included `/controller_epoch` in the "PersistentZkPaths" so that it will be created if not exists on broker startup. In this case, `getControllerEpoch` should not return None unless admin deletes `/controller_epoch` explicitly, which will ruin the cluster anyway.

Reply:
Prior to this patch, we only include znode in ZkData.PersistentZkPaths() if there is no need for the data in the znode. This patch changes this behavior such that we create znode will null data and we assume epoch is -1 if the data is null.

The previous approach says that either the znode does not exist, or the znode exists with valid data. The new approach says that either the znode exists with null, or the znode exists with valid data. I personally prefer the previous approach and I would prefer not to define a znode with null data and add additional code to handle that case. And in general it is probably better to keep the existing code if there is no difference in correctness/performance and the difference in code style is not very obvious w.r.t which one is better.

It looks like the main concern with the previous approach is about code complexity. How about we have create and call method `maybeCreateControllerEpochZnode` at the beginning of `registerController()`?

Reply:
Thanks for the suggestion. I have added `maybeCreateControllerEpochZnode` and avoid crearting the znode on broker start up.